### ISSUE TITLE
Complex-valued layers (e.g: dense)

### ISSUE THREAD

**Issue by PhilipVinc** [google/flax#805]
Hello.
I am one of the main contributors of the [netket](https://github.com/netket/netket) project, a machine-learning for quantum physics package.
Together with other developers, we have migrated most of our code to jax, and I am considering to deeply integrate with flax too.

Note: this issue is a sister issue to [jax#5312](https://github.com/google/jax/issues/5312)

However, in netket we make extended use of complex-valued neural networks, which at the moment are not supported by flax. 
To properly support flax I would like to be able to define layers such as `Dense`, `Conv` and others to use complex-values.
While sometimes we can split the real and complex part of a dense layer by hand, sometimes it would be much easier to have an easier way to do so.

Ideally, I would like the example below to work:

### Problem you have encountered:
Complex-valued `dtype`s with most layers silently return real-valued layers. See for example:

```python
>>> import jax, flax
>>> m=flax.linen.Dense(3, dtype=jax.numpy.complex64)
>>> _, weights = m.init_with_output(jax.random.PRNGKey(0), (3,))
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
>>> weights
FrozenDict({
    params: {
        kernel: DeviceArray([[ 0.67380387, -0.3294223 , -0.9614107 ]], dtype=float32),
        bias: DeviceArray([0., 0., 0.], dtype=float32),
    },
})
```

Where I would have expected a complex-valued network.

the issue probably lies in jax `truncated_normal` (called by `lecun_normal`) not supporting complex values.
I am opening this issue, however, because I am not sure that it would ever make sense for `lecun_normal` to support complex numbers natively, and therefore it might make more sense for flax to have it's own initialisers dispatch to different implementations depending on the required dtype.

In any case, silently returning the wrong output is confusing.

**Comment by jheek** [comment]
I don't think anyone considered complex valued networks before. But I don't see a reason why they should not work.
As far as I can see `jax.random` and `jax.random.initializers` always generate real valued arrays.

In Flax we let the initializer decide the dtype of the paramaters. The dtype of a Module determines only the dtype of the output. This makes sense for most use cases because typically we want float32 weights and some half precision computation (bfloat16/float16).

Here's an example of a Dense network with custom initializers that I think does what you expect.
I defined a complex kernel initializers and the bias is initialized to (complex) zeros.
In this example paramaters, gradients, and activations are in complex64.

```
def complex_kernel_init(rng, shape):
  fan_in = np.prod(shape) // shape[-1]
  x = random.normal(random.PRNGKey(0), shape) + 1j * random.normal(random.PRNGKey(0), shape)
  return x * (2 * fan_in) ** -0.5

complex_bias_init = lambda _, shape: jnp.zeros(shape, jnp.complex64)

m = nn.Dense(3, dtype=jax.numpy.complex64, kernel_init=complex_kernel_init, bias_init=complex_bias_init)
x = jnp.ones((1, 2))
variables = m.init(random.PRNGKey(0), x)

loss_fn = lambda vars: jnp.linalg.norm(m.apply(vars, x)) ** 2

print(m.apply(variables, x))

jax.grad(loss_fn)(variables)
```

**Comment by PhilipVinc** [comment]
Thanks for chiming in.

Indeed Flax works just fine with complex parameters, but that is not my problem: in quantum physics we often mix and match several layers, some of which are only real, some of which have complex weights, and it would be handy if it was easy to only specify a complex `dtype` and let flax dispatch to a different initialiser.

I see now what dtype is supposed to mean, thank you for clarifying that. 
However I'd say that the interface is then a bit... cumbersome if one desires to change the type (even between float32 and float64) of the parameters: instead of simply specifying a kwarg like `param_dtype`, one has to check how many (and what type) of weight initialisers must be passed to every layer, maybe check the default ones, and essentially call again the default ones with a different dtype.

Imagine one wants to create a `float64` version of a Dense layer:
```python
m = nn.Dense(3, dtype=jax.numpy.float64, kernel_init=flax.nn.initializers.lecun_normal(dtype=jax.numpy.float64), bias_init= flax.nn.initializers.normal(dtype=jax.numpy.float64))
``` 
and compare this to a much simpler
```python
m = nn.Dense(3, dtype=jax.numpy.float64, params_dtype=jax.numpy.float64)
```

Moreover, if that was supported, one could also think about adding custom default initialisers in flax that correctly support complex `dtypes`...

**Comment by jheek** [comment]
We might end up making dtype more advanced at one point. One could imagine having a FlaxDType that has a .compute_dtype and .param_dtype (and potentially more). In that case we could prompte simple dtypes to the flax variant whenever a normal dtype is passed to a layer.

On a more pragmatic node: One way to avoid cumbersome repetition of keyword args is to use partial application of Modules. So for example you could define at the top-level module:

```
complex_dense = functools.partial(nn.Dense,  dtype=jax.numpy.complex64, kernel_init=complex_kernel_init, bias_init=complex_bias_init, ...)
real_dense = functools.partial(nn.Dense,  dtype=jax.numpy.float32, ...)
```

and then pass this dense "building block" to lower level modules. Such that in most parts of the model you only have to reason about which building blocks are used instead of the right dtypes and initializers.

**Comment by PhilipVinc** [comment]
I'm still experimenting and i think that at the moment the situation it's quite a bit messy :

The `__call__` code for dense is:
```python
  def __call__(self, inputs: Array) -> Array:
    """Applies a linear transformation to the inputs along the last dimension.
    Args:
      inputs: The nd-array to be transformed.
    Returns:
      The transformed input.
    """
    inputs = jnp.asarray(inputs, self.dtype)
    kernel = self.param('kernel',
                        self.kernel_init,
                        (inputs.shape[-1], self.features))
    kernel = jnp.asarray(kernel, self.dtype)
    y = lax.dot_general(inputs, kernel,
                        (((inputs.ndim - 1,), (0,)), ((), ())),
                        precision=self.precision)
```
 - If I don't change `dtype`, but initialise the parameters to be complex, I get a warning telling me that the imaginary part has been ignored, due to that `jnp.asarray` call. I think that if you explicitly initialised the weights to be complex, you don't want the imaginary part to be stripped away. 
 - If the input of my neural network is complex-valued, but the weights are not, then if `dtype` is real-valued the imaginary part of the input is stripped away. In my use-case, we do not know at network definition time if the inputs will be complex or real.

I understand that dtype is supposed to specify the type used for the calculation, but I propose that it be only the precision of the operation and not the exact dtype: the dtype passed to a dense layer (and similar) should always be real-valued, and, if the inputs or the parameters are complex-valued, the complex-counterpart of the required precision is used.

```python
def complex_dtype(dtype):
    if typ == np.dtype('float32'):
      return np.dtype('complex64')
    elif typ == np.dtype('float64'):
      return np.dtype('complex128')
    else:
      raise TypeError("Complex {} is not supported".format(typ))

def __call__(self, inputs: Array) -> Array:
   if np.issubdtype(inputs.dtype, np.complexfloating) or np.issubdtype(kernel.dtype, np.complexfloating) :
      # it should be enforced somewhere that dtype be real
      dtype = complex_dtype(self.dtype)
   else:
      dtype = self.dtype
   
    inputs = jnp.asarray(inputs, dtype)
    kernel = self.param('kernel',
                        self.kernel_init,
                        (inputs.shape[-1], self.features))
    kernel = jnp.asarray(kernel, dtype)
    y = lax.dot_general(inputs, kernel,
                        (((inputs.ndim - 1,), (0,)), ((), ())),
                        precision=self.precision)
   
```

**Comment by jheek** [comment]
Respecting input dtypes: https://github.com/google/flax/issues/915

I'll leave this issue open for considering more complex dtype promotions when the kernel is complex.

**Comment by PhilipVinc** [comment]
cc @jheek @marcvanzee 
Here are some examples that could be used as tests: https://gist.github.com/PhilipVinc/72fb6d451dc40db4f0cdf15cf585cc3a

I explain things below:

We run tax with `jax_enable_x64=True`.
```python
from jax.config import config
config.update("jax_enable_x64", True)

from jax import random
from jax import numpy as jnp
import flax.linen as nn

key = random.PRNGKey(0)
xr32 = random.normal(key, (4,), dtype=jnp.float32)
xr64 = random.normal(key, (4,), dtype=jnp.float64)
xc32 = random.normal(key, (4,), dtype=jnp.complex64)
xc64 = random.normal(key, (4,), dtype=jnp.complex128)
```

### Inconsistency due to x64

When running with x64 mode enabled, there are some inconsistencies in Flax: initialisers by default switch to `float64` but the `dtype` of flax layers is fixed at `float32`. This is annoying.
```python
dense32 = nn.Dense(kernel_init=nn.initializers.normal(stddev=0.01), dtype=jnp.float32, features=3)
out, pars = dense32.init_with_output(key, xr32)
print(out.dtype) # float32
print(pars['params']['kernel'].dtype) # float64
```

### Problem 1: Type promotion
I would like to treat `ComplexWarning: Casting complex values to real discards the imaginary part
  out = lax._convert_element_type(out, dtype, weak_type=weak_type` as an error. It might help to run CI with that set to error if it is possible.

If the input should be truncated to only real part, the user will do that explicitly in his code. 
I doubt anybody would appreciate a layer implicitly truncating imaginary part, and besides it makes intent hard to read so I would treat it as a bug. This means that flax layers should type-promote

```python
dense64 = nn.Dense(kernel_init=nn.initializers.normal(stddev=0.01), dtype=jnp.float64, features=3)
# I assume that by default this uses float64 for parameters.
# by default it should not cast outputs and just promote

# assuming parameters will be float64, input float32, output will be
# float64 
out, pars = dense64.init_with_output(key, xr32)
assert pars['params']['kernel'].dtype == jnp.float64
assert out.dtype == jnp.float64 

# assuming parameters will be float64, input float64, output will be
# float64 
out, pars = dense64.init_with_output(key, xr64)
assert pars['params']['kernel'].dtype == jnp.float64
assert out.dtype == jnp.float64 # this fails. Why?

# assuming parameters will be float64, input complex32, output will be
# complex128.
out, pars = dense64.init_with_output(key, xc32)
assert pars['params']['kernel'].dtype == jnp.float64
assert out.dtype == jnp.complex128 

out, pars = dense64.init_with_output(key, xc64)
assert pars['params']['kernel'].dtype == jnp.float64
assert out.dtype == jnp.complex128 
```

### Problem 2: quickly changing types of layers
In NetKet we often have to combine layers with complex parameters with layers with real parameters.
If it was easy, we'd be happier.

This is an example of our current API:
```python
dense32 = nn.Dense(dtype=jnp.float32, features=3)
pars = dense32.init(key, xr32)
assert pars['params']['kernel'].dtype == jnp.float32
assert pars['params']['bias'].dtype == jnp.float32

dense64 = nn.Dense(dtype=jnp.float64, features=3)
pars = dense32.init(key, xr32)
assert pars['params']['kernel'].dtype == jnp.float64
assert pars['params']['bias'].dtype == jnp.float64

dense64 = nn.Dense(dtype=jnp.complex64, features=3)
out, pars = dense32.init(key, xr32)
assert pars['params']['kernel'].dtype == jnp.complex64
assert pars['params']['bias'].dtype == jnp.complex64

dense64 = nn.Dense(dtype=jnp.complex128, features=3)
out, pars = dense32.init(key, xr32)
assert pars['params']['kernel'].dtype == jnp.complex128
assert pars['params']['bias'].dtype == jnp.complex128
```

Note: if an user wanted to mix `complex128` kernel and `float64` bias, for example (this is rare, but it shows up in one use-case for Positive-Definite neural networks. Currently I do the bias by hand), one could still do something like

```python
dense32 = nn.Dense(bias_init=lambda k,s,d: normal(stddev=0.01)(k,s), dtype=jnp.float64, features=3)
```

Essentially, we are optimising for the common use case where `layer.dtype == all(parameters.dtype)`, instead of forcing users to specify all dtypes every time.

**Comment by PhilipVinc** [comment]
Hey @jheek, any news on this front?

**Comment by NeilGirdhar** [comment]
@PhilipVinc As an aside, I'm pretty sure I read somewhere that `jax.enablex64` is going to disappear soon?

**Comment by PhilipVinc** [comment]
Yes I'm aware. jax will transition to using float32 by default in most applications, but most of those inconsistencies will still be there.

#1776 solves most of them
